### PR TITLE
Add experiment support

### PR DESCRIPTION
**Pull request by JoshuaKrstic**
Add experiment support.
This change:

- pulls the experiment binary from a GCS bucket (not open-sourceable code)
- gets metadata from the MDS for the experiment binary (image family, image version, git commit hash for presubmit images)
- calls experiment binary with metadata fetched
  - the binary fetches the project id + vm instance ID from the metadata server
- provides experiment data as a map in the launch spec
- provides helper functions to make the launch spec map type-safe

### CODE REVIEW

File: launcher/image/sync_experiments.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
+set -euo pipefail
```
Comment (JoshuaKrstic) [inline launcher/image/sync_experiments.sh:2]: We should consider failing silently given that the experiments server has a lower slo (99%) than us.
We shouldn't block the launcher if experiment sync fails. The experiment class is resilient to the lack of a file + lack of expected flags.

Comment (alexmwu) [reply launcher/image/sync_experiments.sh:2]: This is generally good practice. You otherwise don't have error handling here  for exceptional cases.

If you want to avoid failure, you can wrap it in an if statement and log the error or `|| some_failure_logging_func`. If you think bash is not the best language, consider writing your script in Python or another better scripting language.

File: launcher/image/sync_experiments.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+readonly CS_PATH="${OEM_PATH}/confidential_space"
+readonly EXPERIMENTS_BINARY="confidential_space_experiments"
+readonly SYNC_RESULT_DIRECTORY="/tmp/container_launcher"
+readonly SYNC_RESULT_FILE="experiment_data"
+
+getImageDetails() {
+    # image will look like `projects/confidential-space-images/global/images/confidential-space-hardened-230600`
+    #                   or `projects/confidential-space-images/global/images/confidential-space-debug-230600`
+    #                   or `projects/confidential-space-images/global/images/confidential-space-debug-0-presubmit-abc123`
+    image=$(curl --silent "http://metadata.google.internal/computeMetadata/v1/instance/image" -H "Metadata-Flavor: Google")
```
[the remark this answers is not in the document]
Comment (alexmwu) [reply launcher/image/sync_experiments.sh:14]: That's a capital S, aka `--show-error` in case it's not clear

File: launcher/image/sync_experiments.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+main() {
+    getImageDetails results
+    if [ ! -d "$SYNC_RESULT_DIRECTORY" ]; then
+        echo "$SYNC_RESULT_DIRECTORY does not exist, creating it."
+        mkdir $SYNC_RESULT_DIRECTORY
+    fi
+   
+    echo "calling $CS_PATH/$EXPERIMENTS_BINARY with flags: output=$SYNC_RESULT_DIRECTORY/$SYNC_RESULT_FILE, image_type=\"${results[0]}\" image_version=\"${results[1]}\" -image_build_hash=\"${results[2]}\""
```
[the remark this answers is not in the document]
Comment (alexmwu) [reply launcher/image/sync_experiments.sh:37]: https://pkg.go.dev/os/exec#Cmd.Run. This would make testing easier to implement since we can drop in a mock_experiments_binary.

File: internal/experiments/experiments.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	return readJSONInput(f)
+}
+
+// ReadJSONInput  takes a reader and unmarshalls the contents into the experiments map.
+// If the unmarsahlling fails, the experiments map is set to an empty map.
+func readJSONInput(read io.Reader) (map[string]any, error) {
+	jsondata := make(map[string]any)
+	if err := json.NewDecoder(read).Decode(&jsondata); err != nil {
```
Comment (alexmwu) [inline internal/experiments/experiments.go:92]: I think we should write our own struct if we're gonna be calling something like `GetBoolExperiment` anyway. The default decoder is not strict (although you definitely need a unit test) and will not unmarshal/fail on unknown fields.

Then caller code just inspects `launchSpec.Experiments.TestFeatureForImage`, which is a bool.

File: launcher/image/preload.sh [hunk 5]
```diff
@@ -2,11 +2,21 @@
 
 readonly OEM_PATH='/usr/share/oem'
 readonly CS_PATH="${OEM_PATH}/confidential_space"
+readonly EXPERIMENTS_BINARY="confidential_space_experiments"
+readonly GCS_BUCKET="gs://confidential-space-images_third-party"
```
Comment (alexmwu) [inline launcher/image/preload.sh:6]: This is no longer used

File: launcher/launcher/main.go [hunk 6]
```diff
@@ -28,6 +30,10 @@ const (
 	// panic() returns 2
 	rebootRC = 3 // reboot
 	holdRC   = 4 // hold
+	// hostTokenPath defined the directory in the host that will store attestation tokens and experiment data.
+	hostTokenPath = "/tmp/container_launcher/"
```
Comment (alexmwu) [inline launcher/launcher/main.go:34]: This is duplicated with container_runner.go. Any reason we don't pull it out into some constants or common package

File: launcher/image/sync_experiments.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (JoshuaKrstic) [reply launcher/image/sync_experiments.sh:14]: This will be removed and put into the binary instead.

File: internal/experiments/experiments.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	return f, true
+}
+
+// ReadExperimentsFile takes a filepath, opens the file, and calls ReadJsonInput with the contents
+// of the file.
+// If the file cannot be opened, the experiments map is set to an empty map.
+func ReadExperimentsFile(fpath string) (map[string]any, error) {
+	f, err := os.Open(fpath)
```
[the remark this answers is not in the document]
Comment (JoshuaKrstic) [reply internal/experiments/experiments.go:80]: thanks for catching this.

File: launcher/launcher/main.go [hunk 8]
```diff
@@ -72,6 +78,13 @@ func main() {
 		return
 	}
 
+	e, err := experiments.ReadExperimentsFile(path.Join(hostTokenPath, experimentDataFile))
+	if err != nil {
+		logger.Println(err)
```
[the remark this answers is not in the document]
Comment (JoshuaKrstic) [reply launcher/launcher/main.go:83]: done.

File: launcher/image/sync_experiments.sh (hunk 1 above)
Comment (JoshuaKrstic) [reply launcher/image/sync_experiments.sh:2]: This script has been removed entirely.

File: launcher/image/sync_experiments.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (JoshuaKrstic) [reply launcher/image/sync_experiments.sh:16]: Done

File: internal/experiments/experiments.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+	"io"
+	"os"
+)
+
+// GetBooleanExperiment attempts to read the key from the map provided and does type checking on the
+// returned value, if any. Value is false on failure.
+func GetBooleanExperiment(e map[string]any, key string) (bool, bool) {
+	v, err := e[key]
+	if !err {
+		return false, false
```
[the remark this answers is not in the document]
Comment (JoshuaKrstic) [reply internal/experiments/experiments.go]: Done.

File: internal/experiments/experiments.go (hunk 4 above)
Comment (JoshuaKrstic) [reply internal/experiments/experiments.go]: Removed in favor of generics.

File: internal/experiments/experiments_test.go [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+	e, err := readJSONInput(&buffer)
+	if err != nil {
+		t.Errorf("Failed to parse input json, err: %v", err)
+	}
+
+	val, ok := GetStringExperiment(e, "FloatFeature")
+	if ok {
+		t.Errorf("expected to fail to get string experiment, got %v", val)
+	}
+}
```
[the remark this answers is not in the document]
Comment (JoshuaKrstic) [reply internal/experiments/experiments_test.go]: Love these, thanks for the tip.

File: launcher/launcher/main.go (hunk 6 above)
Comment (alexmwu) [reply launcher/launcher/main.go]: please rebase off of https://github.com/google/go-tpm-tools/pull/356

File: internal/experiments/experiments.go (hunk 4 above)
Comment (alexmwu) [reply internal/experiments/experiments.go]: I think this is a better approach since the type checking is in one place instead of spread around. Can we do this instead of generics? Sorry if I was unclear; I meant to remove the generics comments

Comment (JoshuaKrstic) [reply internal/experiments/experiments.go]: Gotcha, that certainly clears things up. Made the change.

File: launcher/internal/experiments/experiments.go [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+	return r, err
+}
+
+// ReadJSONInput  takes a reader and unmarshalls the contents into the experiments map.
+// If the unmarsahlling fails, the experiments map is set to an empty map.
+func readJSONInput(b []byte) (Experiments, error) {
+	var experiments Experiments
+	if err := json.Unmarshal(b, &experiments); err != nil {
+		// Return default values on failure.
+		return Experiments{}, fmt.Errorf("failed to unmarshall json: %w", err)
```
Comment (alexmwu) [inline launcher/internal/experiments/experiments.go:39]: nit: here and elsewhere "unmarshall" -> "unmarshal"

File: launcher/image/preload.sh [hunk 12]
```diff
@@ -2,11 +2,18 @@
...
 readonly OEM_PATH='/usr/share/oem'
 readonly CS_PATH="${OEM_PATH}/confidential_space"
+readonly EXPERIMENTS_BINARY="confidential_space_experiments"
 
 copy_launcher() {
   cp launcher "${CS_PATH}/cs_container_launcher"
 }
 
+copy_experiment_client() {
+  # EXPERIMENTS_BINARY is placed at  by the `gcloud` step in image/cloudbuild.yaml
```
Comment (alexmwu) [inline launcher/image/preload.sh:12]: nit: DownloadExpBinary creates the file at EXPERIMENTS_BINARY.

File: launcher/internal/experiments/experiments_test.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+import (
+	"testing"
+)
+
+func TestExperiments(t *testing.T) {
+	tests := []struct {
+		input string
+	}{
+		{input: "{\"EnableTestFeatureForImage\":true,\"EnableSignedContainerImage\":true}"},
```
Comment (alexmwu) [inline launcher/internal/experiments/experiments_test.go:11]: Add a test for invalid JSON (e.g., missing a closing bracket).

File: launcher/internal/experiments/experiments_test.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
+package experiments
+
+import (
+	"testing"
+)
+
```
Comment (alexmwu) [inline launcher/internal/experiments/experiments_test.go:6]: Add a test for empty string and empty JSON "{}".

File: launcher/launcher/main.go [hunk 15]
```diff
@@ -72,6 +79,16 @@ func main() {
 		return
 	}
 
+	args := fmt.Sprintf("-output=%s", experimentDataFile)
+	exec.Command(binaryPath, args).Run()
```
Comment (alexmwu) [inline launcher/launcher/main.go:83]: You should also catch the err from Run()

```
if err != nil {
    logger.Printf("failed to fetch experiments: %v\n", err)
}
```

File: launcher/launcher/main.go [hunk 16]
```diff
@@ -72,6 +79,16 @@ func main() {
...
+
+	e, err := experiments.New(path.Join(launcherfile.HostTmpPath, experimentDataFile))
+	if err != nil {
+		logger.Println("failed to read experiment file %w", err)
```
Comment (alexmwu) [inline launcher/launcher/main.go:87]: If you are using the a format verb, you need to use logger.Printf. The format verb %w is typically used when returning an error however, so you don't need to wrap the err. Just use %v or:
`logger.Println("Failed to read experiment file: ", err)`

File: launcher/launcher/main.go [hunk 17]
```diff
@@ -72,6 +79,16 @@ func main() {
 		return
 	}
 
+	args := fmt.Sprintf("-output=%s", experimentDataFile)
```
Comment (alexmwu) [inline launcher/launcher/main.go:82]: Does output assume a prefix of `launcherfile.HostTmpPath`? Or do you need to `path.Join` with `experimentDataFile` here? (if so, you can pull out the path.Join below into a variable)

**Review by alexmwu — CHANGES_REQUESTED** [review]

File: launcher/internal/experiments/experiments_test.go (hunk 13 above)
Comment (JoshuaKrstic) [reply launcher/internal/experiments/experiments_test.go:11]: This is in the next test case.

File: launcher/internal/experiments/experiments_test.go (hunk 14 above)
Comment (JoshuaKrstic) [reply launcher/internal/experiments/experiments_test.go:6]: Added the empty string, the empty bracket is in the next test case.

File: launcher/launcher/main.go (hunk 16 above)
Comment (JoshuaKrstic) [reply launcher/launcher/main.go:87]: Gotcha, thanks for the tip. I think the println is an artifact from previous statement, thanks for catching that,

File: launcher/launcher/main.go (hunk 17 above)
Comment (JoshuaKrstic) [reply launcher/launcher/main.go:82]: oopsie, it does need the combo. I appear to have rushed this commit :)

**Review by alexmwu — APPROVED** [review]
